### PR TITLE
add ability to filter metrics on a per-forwarder basis

### PR DESCRIPTION
**Pull request by mdubbyap**

### CODE REVIEW

**Comment by mdubbyap** [comment]
@mpetazzoni  also

File: protocol/signalfx/signalfxforwarder.go [hunk 1]
```diff
@@ -59,6 +62,7 @@ type ForwarderConfig struct {
 }
 
 var defaultForwarderConfig = &ForwarderConfig{
+	Filters:      filtering.FilterObj{}.Pointer(),
```
Comment (cep21) [inline protocol/signalfx/signalfxforwarder.go]: It's just built ins you can't take literals to, like `&"hello"`.  You can do `&filtering.FilterObj{}`.

File: protocol/filtering/forwarder.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+			break
+		}
+	}
+	if found || (len(f.allow) == 0 && !denied) {
+		return true
+	}
+	return false
+}
+
+// FilterDatapoints filters datapoints based on the metric name as counts how many it filters
```
Comment (mpetazzoni) [inline protocol/filtering/forwarder.go]: "as well as"

File: protocol/filtering/forwarder.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	}
+	found := false
+	for _, a := range f.allow {
+		if a.Match([]byte(metricName)) {
+			denied = false
+			found = true
+			break
+		}
+	}
+	if found || (len(f.allow) == 0 && !denied) {
```
Comment (mpetazzoni) [inline protocol/filtering/forwarder.go]: `return found || ...` directly?

File: protocol/filtering/forwarder.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+		if f.FilterMetricName(d.Metric) {
+			validDatapoints = append(validDatapoints, d)
+		}
+	}
+	invalidDatapoints := len(datapoints) - len(validDatapoints)
+	atomic.AddInt64(&f.FilteredDatapoints, int64(invalidDatapoints))
+	return validDatapoints
+}
+
+// GetFilteredDatapoints returns a cumulative counter of how many filtered datapoints by this forwarder
```
Comment (mpetazzoni) [inline protocol/filtering/forwarder.go]: grammar

File: README.md [hunk 5]
```diff
@@ -613,6 +613,47 @@ points to signalfx.
...
+  "StatsDelay": "1s",
+  "ListenFrom": [
+    {
+      "Type": "carbon",
+      "ListenAddr" : "0.0.0.0:2003",
+      "MetricDeconstructor": "commakeys"
+    },
+    {
+      "Type": "signalfx",
+      "ListenAddr" : "0.0.0.0:8080",
```
Comment (mpetazzoni) [inline README.md]: Pretty sure that trailing comma shouldn't be there

File: README.md [hunk 6]
```diff
@@ -613,6 +613,47 @@ points to signalfx.
...
+    },
```
Comment (mpetazzoni) [inline README.md]: Ditto

File: README.md [hunk 7]
```diff
@@ -613,6 +613,47 @@ points to signalfx.
...
+  ],
+
+  "ForwardTo": [
+    {
+      "type": "carbon",
+      "Name": "ourcarbon",
+      "Host": "example.com",
+      "Port": 2003,
```
Comment (mpetazzoni) [inline README.md]: Ditto

File: README.md [hunk 8]
```diff
@@ -613,6 +613,47 @@ points to signalfx.
...
+would be denied and those that were not would be allowed.
+
+```
+{
+  "StatsDelay": "1s",
+  "ListenFrom": [
+    {
+      "Type": "carbon",
+      "ListenAddr" : "0.0.0.0:2003",
+      "MetricDeconstructor": "commakeys"
```
Comment (mpetazzoni) [inline README.md]: Is this required?

File: .gitignore [hunk 9]
```diff
@@ -3,3 +3,4 @@ metricproxy.pid
 /*.iml
 /sfproxytest.json
 /metricproxy
+/sf
```
Comment (mpetazzoni) [inline .gitignore]: ?

File: README.md (hunk 8 above)
Comment (mdubbyap) [inline README.md]: no, there's a default, just being explicit

File: .gitignore (hunk 9 above)
Comment (mdubbyap) [inline .gitignore]: default directory for proxy configs, i was tired of having untracked files

**Pull request merged**